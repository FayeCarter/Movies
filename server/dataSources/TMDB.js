import { RESTDataSource } from 'apollo-datasource-rest';
require("dotenv").config();

export class TMDBAPI extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = "https://api.themoviedb.org/3/";
    this.authKey = process.env.TMDB_AUTH;
  }

  movieReducer({ id, title, poster_path, overview, genre_ids, vote_average } = {}) {
    return {
      id,
      title,
      image: poster_path,
      about: overview,
      genre_ids,
      rating: vote_average,
    }
  }

  genreReducer({ id, name } = {}) {
    return {
      id,
      genre: name,
    }
  }

  async getPopularMovies() {
    const result = await this.get(`movie/popular?page={0}&api_key=${this.authKey}`);
    return result?.results?.map(movie => this.movieReducer(movie));
  }

  async getGenres() {
    const result = await this.get(`genre/list?api_key=${this.authKey}`);
    return result?.genres.map(genre => this.genreReducer(genre));
  }

}

