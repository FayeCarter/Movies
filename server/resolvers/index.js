export const resolvers = {
  Query: {
    popularMovies: (_, __, { dataSources }) =>
    Promise.all(
      Object.keys(dataSources).map(source => 
        dataSources[source].getPopularMovies()
      )
    ).then(result => result.reduce((acc, data) =>  acc.concat(data), [])),

    movieGenres: (_, __, { dataSources }) =>
      Promise.all(
        Object.keys(dataSources).map(source => 
          dataSources[source].getGenres()
        )
      ).then(result => result.reduce((acc, data) =>  acc.concat(data), [])),
  }
}