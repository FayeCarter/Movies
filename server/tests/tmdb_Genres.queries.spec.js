import { createTestClient } from 'apollo-server-testing';
import { ApolloServer, gql } from 'apollo-server';
import { TMDBAPI } from "../datasources/TMDB";
import { resolvers } from '../resolvers';
import { typeDefs } from '../typedefs';
import { getGenrePreReducerStub } from '../fixtures/TMDB_Genres';

const GET_MOVIE_GENRES= gql`
  query getGenres{
    movieGenres {
      id
      genre
    }
  }
`;

const constructTestServer = () => {
  const tmdbAPI = new TMDBAPI();

  const server = new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => ({
      tmdb: tmdbAPI
    })
  });

  return { server, tmdbAPI };
}

describe('[Queries.TMDBAPI]', () => {
  it('fetches an array of genres from the TMDB API', async () => {
    const { server, tmdbAPI } = constructTestServer();

    tmdbAPI.get = jest.fn(() => ({
      results: [getGenrePreReducerStub],
    }));

    const { query } = createTestClient(server);

    const response = await query({
      query: GET_MOVIE_GENRES,
      variables: { ids: [28] },
    });

    expect(response).toMatchSnapshot();
  });
});