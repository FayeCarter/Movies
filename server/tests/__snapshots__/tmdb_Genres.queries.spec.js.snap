// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Queries.TMDBAPI] fetches an array of genres from the TMDB API 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Cannot read property 'map' of undefined],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
